### PR TITLE
Add support for csrf tokens in html forms with jinja

### PR DESCRIPTION
**Pull request by LainezDev**
Now FastAPI with support for csrf tokens in html forms with jinja

Added some improvements for using CSRF tokens in HTML forms without affecting the current functionality of the middleware **starlette-csrf**. The improvement simplifies the process of obtaining the CSRF token by unifying the search logic into a single function _get_submitted_csrf_token. Now, the function will first attempt to obtain the token from the headers, and if it does not find it there, it will search for it in the form. Additionally, this new functionality addresses the crash issue that occurs in Starlette due to body consumption.

This is particularly useful for integration with FastAPI and the development of secure websites utilizing forms.

Now, all that's needed is to add the starlette_csrf middleware and utilize the following template processor in your FastAPI code:

```py
from fastapi import FastAPI
from fastapi.templating import Jinja2Templates
from fastapi.middleware.csrf import CSRFMiddleware, csrf_token_processor

app = FastAPI()

your_cookie_name = "your_cookie_name"
your_header_name = "your_header_name"

app.add_middleware(
   CSRFMiddleware, 
   secret = "your_secret",
   cookie_name = your_cookie_name,
   header_name = your_header_name,
   )

templates = Jinja2Templates(
    directory="templates", 
    context_processors=[csrf_token_processor(your_cookie_name, your_header_name)]
    )

```
Now, the middleware integrates with a context processor, a function that returns a dictionary containing the CSRF token, CSRF input, and CSRF header for use with other tools such as HTMX.

Simply using {{ csrf_input | safe }} in each form is now sufficient to ensure a more secure web application. For example:

```html
<form method="post">
    {{ csrf_input | safe }}
    <!-- Other form fields here -->
    <button type="submit">Submit</button>
</form>
```

Furthermore, we can use {{ csrf_header }} in HTMX requests. For example:

```html
<form hx-patch="/route/edit" hx-headers='{{ csrf_header | tojson | safe }}'  hx-trigger="submit" hx-target="#yourtarget" hx-swap="outerHTML" >
    <!-- Other form fields here -->
    <button type="submit">Submit</button>
</form>
```

Feel free to let me know if you need further adjustments or improvements!

### CODE REVIEW

File: fastapi/middleware/csrf.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            and self._has_sensitive_cookies(request.cookies)
+        ):
+            submitted_csrf_token = await self._get_submitted_csrf_token(request)
+
+            if (
+                not csrf_cookie
+                or not submitted_csrf_token
+                or not self._csrf_tokens_match(csrf_cookie, submitted_csrf_token)
+            ):
+                response = self._get_error_response(request)
```
Comment (GDemay) [inline fastapi/middleware/csrf.py:70]:
```suggestion
                response = self._get_error_response(request=request)
```

File: fastapi/middleware/csrf.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                await response(scope, receive, send)
+                return
+
+        request._receive = self._receive_with_body(request._receive, body)
```
Comment (GDemay) [inline fastapi/middleware/csrf.py:74]: Missing typing

File: fastapi/middleware/csrf.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    def _url_is_exempt(self, url: URL) -> bool:
+        if not self.exempt_urls:
+            return False
+        for exempt_url in self.exempt_urls:
+            if exempt_url.match(url.path):
+                return True
+        return False
+
+    async def _get_request_body(self, request: Request):
+        if request.method in ("POST", "PUT", "PATCH", "DELETE"):
```
Comment (GDemay) [inline fastapi/middleware/csrf.py:123]: What do you think about adding it in a settings file?

```
from enum import Enum
from pydantic import BaseSettings

class HttpMethod(str, Enum):
    POST = "POST"
    PUT = "PUT"
    PATCH = "PATCH"
    DELETE = "DELETE"


class Settings(BaseSettings):
    allowed_methods: list[HttpMethod] = [
        HttpMethod.POST, 
        HttpMethod.PUT, 
        HttpMethod.PATCH, 
        HttpMethod.DELETE
    ]

settings = Settings()

```

File: fastapi/middleware/csrf.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    def _generate_csrf_token(self) -> str:
+        token = self.serializer.dumps(secrets.token_urlsafe(128))
+        return cast(str, token)
+
+    def _csrf_tokens_match(self, token1: str, token2: str) -> bool:
+        try:
+            decoded1: str = self.serializer.loads(token1)
+            decoded2: str = self.serializer.loads(token2)
+            return secrets.compare_digest(decoded1, decoded2)
+        except BadSignature:
```
Comment (GDemay) [inline fastapi/middleware/csrf.py:148]: Should we had a logger here?

**Review by GDemay — APPROVED** [review]
Good work!

**Review by patrick91 — COMMENTED** [review]
This seems to me very similar to https://github.com/frankie567/starlette-csrf/blob/main/starlette_csrf/middleware.py with some additions

@LainezDev maybe it is worth adding those addition directly to starlette_csrf? 😊

**Comment by tiangolo** [comment]
Thanks for the interest @LainezDev!

Yep, as @patrick91 says, probably better to contribute the changes to that repo. :nerd_face: 

I'll pass on this one here, but thanks for the effort! :coffee:

**Pull request closed without merging**